### PR TITLE
Implement variable fixing for forms

### LINKED ISSUE #99: fix_variables is an empty stub (silent no-op) — implement variable fixing or remove it
Part of the benchmarking-readiness epic: JuliaQUBO/QUBO.jl#40 (WS3).

## Problem

`fix_variables` in `src/library/form/abstract.jl:166-173` is an **empty stub** — the loop body is blank, the function computes nothing and returns `nothing` implicitly:

```julia
function fix_variables(fix::Vector{<:Integer}, Φ::F) where {T,F<:AbstractForm{T}}
    n = dimension(Φ)
    L = linear_form(Φ)

    for i in fix

    end
end
```

Any caller gets a silent no-op. For the benchmarking campaign, variable fixing is the standard preprocessing step (persistencies, warm-start reduction), and a silently-broken primitive is worse than a missing one.

## Proposed PR

Either **implement** or **remove** — do not leave the stub. Recommended: implement, since preprocessing large QOBLib instances will want it.

Implementation sketch (signature change suggested — fixing needs *values*, not just indices):

```julia
fix_variables(Φ::AbstractForm{T}, fix::AbstractDict{Int,U}) -> (Φ′, offset_delta, index_map)
```

- For each fixed variable `i` with value `xᵢ`: fold `Lᵢ·xᵢ` into the offset; for each quadratic term `Q_{ij}`, fold `Q_{ij}·xᵢ` into `L_j` (or into the offset when both endpoints fixed).
- Reindex the surviving variables densely; return the old→new `index_map` so solutions can be lifted back.
- Domain handling: define semantics for both `𝔹` (values in {0,1}) and `𝕊` (values in {−1,+1}); validate inputs against `domain(Φ)`.
- Implement generically over `AbstractForm` via the `linear_terms`/`quadratic_terms` iterators, with a specialized fast path for the sparse forms (operate on `findnz` output directly).
- Add `lift_state(state_reduced, fix, index_map, n)` to reconstruct full states — needed by anything evaluating fixed-reduced solutions against the original model.

## Tests (required in the PR)

- Round-trip property test: for random models (both domains, dense and sparse forms, n ≤ 64), `value(Φ, x) == value(Φ′, x_reduced) (+ scale/offset accounting)` for all assignments consistent with the fixing.
- Edge cases: fix all variables (result is pure offset); fix none (identity); fixing an index out of range throws.

## Checkpoints (anti-drift)

- [ ] **Before implementation**: decide implement-vs-remove and the final signature on this issue (the dict-of-values signature above breaks the current stub's signature — fine, since the stub is unusable, but say so in NEWS.md).
- [ ] **Mid-PR**: post the round-trip property test passing for all four form types × two domains.
- [ ] **Before merge**: grep the ecosystem (ToQUBO.jl, QUBODrivers.jl, QUBOLib.jl) for `fix_variables` callers; confirm none rely on the old no-op signature.

### PR DESCRIPTION
**Pull request by bernalde**
## Summary

- Implements `fix_variables(form, fix::AbstractDict)` for `AbstractForm` backends by folding fixed linear and quadratic contributions into the reduced form offset and surviving linear terms.
- Adds `lift_state` to reconstruct full states from reduced states, fixed values, and the returned old-to-new index map.
- Documents the breaking signature change in the changelog and lists the new API in the docs.

Closes #99

## Tests

- `julia +1.12 --project=. -e 'using Test, Random, SparseArrays, QUBOTools; include("test/assets/comparison.jl"); include("test/unit/library/form.jl"); test_form_fix_variables()'` passed: 390 variable-fixing tests.
- `julia +1.12 --project=. -e 'using Pkg; Pkg.test()'` passed: 1105 tests.

Local note: the default `julia` launcher is Julia 1.10, but the checked-in manifest is resolved with Julia 1.12; local validation used Julia 1.12. CI still covers the configured Julia 1.10 and latest-stable matrix.

## Downstream Search

- `gh search code fix_variables --repo JuliaQUBO/ToQUBO.jl --json path,repository,url --limit 100`: no references.
- `gh search code fix_variables --repo JuliaQUBO/QUBODrivers.jl --json path,repository,url --limit 100`: no references.
- `gh search code fix_variables --repo JuliaQUBO/QUBOLib.jl --json path,repository,url --limit 100`: no references.

## Branch Hygiene

- Base branch: `main`
- Source branch point: `origin/main` at `a3637a0`
- Stacked status: not stacked
- Prerequisite PRs: none

### CODE REVIEW

File: src/library/form/form.jl [hunk 1]
```diff
@@ -69,6 +69,181 @@ function Base.copy(
...
+    for (i, j) in index_map
+        1 <= i <= n || throw(ArgumentError("index_map key $i is out of range 1:$n"))
+        !haskey(fixed, i) || throw(ArgumentError("index_map contains fixed variable $i"))
+        1 <= j <= m || throw(ArgumentError("reduced variable index $j is out of range 1:$m"))
+        j ∉ seen || throw(ArgumentError("index_map contains duplicate reduced index $j"))
+
+        push!(seen, j)
+    end
+
+    length(fixed) + length(index_map) == n ||
```
Comment (bernalde) [inline src/library/form/form.jl:244]: Blocking: this validation only checks that mapped indices are in bounds and non-duplicated; it does not check that every entry of `state_reduced` is actually used. For example, `lift_state([0, 1], Dict{Int,Int}(), Dict(1 => 1), 1)` returns `[0]` and silently ignores the extra reduced value. Please require `length(index_map) == length(state_reduced)` or otherwise verify that `seen` covers `1:m`, and add a regression test for too-long reduced states.

**Review by bernalde — COMMENTED** [review]
Blocking issues:

- Blocking: `lift_state` accepts a reduced state that is longer than the index map and silently drops the extra entries. That can let callers evaluate or store a lifted solution that was not the reduced solution they passed in. I reproduced this on the PR head with `QUBOTools.lift_state([0, 1], Dict{Int,Int}(), Dict(1 => 1), 1)`, which returns `[0]` instead of throwing. Require the reduced state length to match the number of surviving variables, and add a regression test for the too-long case.

Nonblocking issues: none.

Questions: none.

Tests run and outcomes:

- Confirmed GitHub and local diff totals match: 7 files changed, 314 insertions, 9 deletions.
- Read linked issue #99 and verified this PR targets the requested `fix_variables` implementation and state lifting API.
- `julia +1.12 --project=. -e 'using Test, Random, SparseArrays, QUBOTools; include("test/assets/comparison.jl"); include("test/unit/library/form.jl"); test_form_fix_variables()'`: passed, 390 tests.
- `julia +1.12 --project=. -e 'using Pkg; Pkg.test()'`: passed, 1105 tests. The local test environment re-resolved in `/tmp` because the test manifest differs from the root manifest.
- Manual reproduction: `QUBOTools.lift_state([0, 1], Dict{Int,Int}(), Dict(1 => 1), 1)` returned `[0]`.

Merge-readiness: I would not merge this until the blocking issue above is addressed. Because I am posting from the PR author's GitHub account, GitHub requires this review to be submitted as COMMENT; the formal approval or change request must come from another maintainer.

**Comment by bernalde** [comment]
Addressed the unresolved review thread.

Commits pushed:

- `4b165da` - `fix: validate reduced state length (#102 review)`

Main changes made:

- `lift_state` now rejects reduced states whose length does not match the returned `index_map`.
- Added a regression test for too-long reduced states across all variable-fixing form/domain cases.

Tests run and results:

- Confirmed the new regression test failed before the validator change.
- `julia +1.12 --project=. -e 'using Test, Random, SparseArrays, QUBOTools; include("test/assets/comparison.jl"); include("test/unit/library/form.jl"); test_form_fix_variables()'`: passed, 396 tests.
- Manual reproduction now throws `ArgumentError`: `QUBOTools.lift_state([0, 1], Dict{Int,Int}(), Dict(1 => 1), 1)`.
- `git diff --check`: passed.
- `julia +1.12 --project=. -e 'using Pkg; Pkg.test()'`: passed, 1111 tests.

Comments intentionally not addressed:

- The automated benchmark report was treated as informational, not a review request. No code change was made for it.

Remaining risks or follow-up:

- No unresolved blocking review comments were declined.
- New CI jobs from this push are still pending; local validation is complete.

File: src/library/form/form.jl (hunk 1 above)
Comment (bernalde) [reply src/library/form/form.jl:244]: Addressed in https://github.com/JuliaQUBO/QUBOTools.jl/commit/4b165da. `_lift_state_validate` now requires the reduced state length to match `index_map`, and `test_form_fix_variables()` covers the too-long reduced-state case. Focused tests now pass with 396 checks; full `Pkg.test()` passes with 1111 checks.
